src/backend/distributed/utils/placement_connection.c

```diff
@@ -0,0 +1,574 @@
+ * invalid, and/or whether sufficiently many placements have failed to abort
+ * the entire coordinated transaction.
+ *
+ * This will usually be called twice. Once before the remote commit is done,
+ * and once after. This is so we can abort before executing remote commits,
+ * and so we can handle remote transactions that failed during commit.
+ *
+ * When preCommit or using2PC is true, failures on transactions marked as
+ * critical will abort the entire coordinated transaction. Otherwise we can't
+ * anymore, because some remote transactions might have already committed.
+ */
+void
+CheckForFailedPlacements(bool preCommit, bool using2PC)
+{
+	HASH_SEQ_STATUS status;
+	ConnectionShardHashEntry *shardEntry = NULL;
+	int successes = 0;
+	int attempts = 0;
+
+	hash_seq_init(&status, ConnectionShardHash);
+	while ((shardEntry = (ConnectionShardHashEntry *) hash_seq_search(&status)) != 0)
+	{
+		attempts++;
+		if (CheckShardPlacements(shardEntry, preCommit, using2PC))
+		{
+			successes++;
+		}
+	}
+
+	/*
+	 * If no shards could be modified at all, error out. Doesn't matter if
+	 * we're post-commit - there's nothing to invalidate.
+	 */
+	if (attempts > 0 && successes == 0)
+	{
+		ereport(ERROR, (errmsg("could not commit transaction on any active node")));
+	}
+}
+
+
+/*
+ * CheckShardPlacements is a helper function for CheckForFailedPlacements that
+ * performs the per-shard work.
+ */
+static bool
+CheckShardPlacements(ConnectionShardHashEntry *shardEntry,
+					 bool preCommit, bool using2PC)
+{
+	int failures = 0;
+	int successes = 0;
+	dlist_iter placementIter;
+
+	dlist_foreach(placementIter, &shardEntry->placementConnections)
+	{
+		ConnectionPlacementHashEntry *placementEntry =
+			dlist_container(ConnectionPlacementHashEntry, shardNode, placementIter.cur);
+		dlist_iter referenceIter;
+
+		dlist_foreach(referenceIter, &placementEntry->connectionReferences)
+		{
+			ConnectionReference *reference =
+				dlist_container(ConnectionReference, placementNode, referenceIter.cur);
+			MultiConnection *connection = reference->connection;
+
+			/*
+			 * If neither DDL nor DML were executed, there's no need for
+			 * invalidation.
+			 */
+			if (!reference->hadDDL && !reference->hadDML)
+			{
+				continue;
+			}
+
+			/*
+			 * Failed if connection was closed, or remote transaction was
+			 * unsuccessful.
+			 */
+			if (!connection || connection->remoteTransaction.transactionFailed)
+			{
+				placementEntry->failed = true;
+			}
+		}
+
+		if (placementEntry->failed)
+		{
+			failures++;
+		}
+		else
+		{
+			successes++;
+		}
+	}
+
+	if (failures > 0 && successes == 0)
+	{
+		int elevel = 0;
+
+		/*
+		 * Only error out if we're pre-commit or using 2PC. Otherwise we can
+		 * end up with a state where parts of the transaction is committed and
+		 * others aren't, without correspondingly marking things as invalid
+		 * (which we can't, as we would have already committed).  This sucks.
+		 */
+		if (preCommit || using2PC)
```

Comment:
We can't error out, because there might be other shards that have been [partially] modified and committed successfully :( (if the other shard had fewer placements).  If no modifications at all are successful we'll error out later in CheckForFailedPlacements().   I'll add a comment explaining that we'll fail later if no modifications were successful?